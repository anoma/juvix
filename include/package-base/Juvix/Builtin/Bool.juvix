module Juvix.Builtin.Bool;

--- Inductive definition of booleans.
builtin bool
type Bool :=
  | true
  | false;
