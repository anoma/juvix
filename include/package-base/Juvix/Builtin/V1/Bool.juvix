module Juvix.Builtin.V1.Bool;

--- Inductive definition of booleans.
builtin bool
type Bool :=
  | false
  | true;

open Bool using {false; true} public;
