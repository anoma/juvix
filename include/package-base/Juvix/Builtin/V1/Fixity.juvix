module Juvix.Builtin.V1.Fixity;

syntax fixity none := none;

syntax fixity rapp := binary {assoc := right};
syntax fixity lapp := binary {assoc := left; same := rapp};
syntax fixity seq := binary {assoc := left; above := [lapp]};

syntax fixity functor := binary {assoc := right};

syntax fixity logical := binary {assoc := right; above := [seq]};
syntax fixity comparison := binary {assoc := none; above := [logical]};

syntax fixity pair := binary {assoc := right};
syntax fixity cons := binary {assoc := right; above := [pair]};

syntax fixity step := binary {assoc := right};
syntax fixity range := binary {assoc := right; above := [step]};

syntax fixity additive := binary {assoc := left; above := [comparison; range; cons]};
syntax fixity multiplicative := binary {assoc := left; above := [additive]};

syntax fixity composition := binary {assoc := right; above := [multiplicative]};
syntax fixity lcomposition := binary {assoc := left; above := [multiplicative]};
