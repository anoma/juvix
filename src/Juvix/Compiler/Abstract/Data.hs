module Juvix.Compiler.Abstract.Data
  ( module Juvix.Compiler.Abstract.Data.Name,
    module Juvix.Compiler.Abstract.Data.InfoTable,
    module Juvix.Compiler.Abstract.Data.InfoTableBuilder,
  )
where

import Juvix.Compiler.Abstract.Data.InfoTable
import Juvix.Compiler.Abstract.Data.InfoTableBuilder
import Juvix.Compiler.Abstract.Data.Name
