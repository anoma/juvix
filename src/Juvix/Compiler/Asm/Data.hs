module Juvix.Compiler.Asm.Data
  ( module Juvix.Compiler.Asm.Data.InfoTable,
    module Juvix.Compiler.Asm.Data.InfoTableBuilder,
    module Juvix.Compiler.Asm.Data.Stack,
  )
where

import Juvix.Compiler.Asm.Data.InfoTable
import Juvix.Compiler.Asm.Data.InfoTableBuilder
import Juvix.Compiler.Asm.Data.Stack hiding (empty)
