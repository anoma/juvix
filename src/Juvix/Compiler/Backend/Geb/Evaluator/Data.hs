module Juvix.Compiler.Backend.Geb.Evaluator.Data
  ( module Juvix.Compiler.Backend.Geb.Evaluator.Data.Env,
    module Juvix.Compiler.Backend.Geb.Evaluator.Data.Values,
    module Juvix.Compiler.Backend.Geb.Evaluator.Data.RunEvalResult,
  )
where

import Juvix.Compiler.Backend.Geb.Evaluator.Data.Env
import Juvix.Compiler.Backend.Geb.Evaluator.Data.RunEvalResult
import Juvix.Compiler.Backend.Geb.Evaluator.Data.Values
