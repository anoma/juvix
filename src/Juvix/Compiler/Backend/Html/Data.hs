module Juvix.Compiler.Backend.Html.Data (module Juvix.Compiler.Backend.Html.Data.Options) where

import Juvix.Compiler.Backend.Html.Data.Options
