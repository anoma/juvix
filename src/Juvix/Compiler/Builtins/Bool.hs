module Juvix.Compiler.Builtins.Bool where

import Juvix.Compiler.Abstract.Extra
import Juvix.Compiler.Abstract.Pretty
import Juvix.Compiler.Builtins.Effect
import Juvix.Prelude

registerBoolDef :: (Member Builtins r) => InductiveDef -> Sem r ()
registerBoolDef d = do
  unless (null (d ^. inductiveParameters)) (error "Bool should have no type parameters")
  unless (isSmallUniverse' (d ^. inductiveType)) (error "Bool should be in the small universe")
  registerBuiltin BuiltinBool (d ^. inductiveName)
  case d ^. inductiveConstructors of
    [c1, c2] -> registerTrue c1 >> registerFalse c2
    _ -> error "Bool should have exactly two constructors"

registerTrue :: (Member Builtins r) => InductiveConstructorDef -> Sem r ()
registerTrue d@InductiveConstructorDef {..} = do
  let ctorTrue = _constructorName
      ctorTy = _constructorType
  boolTy <- getBuiltinName (getLoc d) BuiltinBool
  unless (ctorTy === boolTy) (error $ "true has the wrong type " <> ppTrace ctorTy <> " | " <> ppTrace boolTy)
  registerBuiltin BuiltinBoolTrue ctorTrue

registerFalse :: (Member Builtins r) => InductiveConstructorDef -> Sem r ()
registerFalse d@InductiveConstructorDef {..} = do
  let ctorFalse = _constructorName
      ctorTy = _constructorType
  boolTy <- getBuiltinName (getLoc d) BuiltinBool
  unless (ctorTy === boolTy) (error $ "false has the wrong type " <> ppTrace ctorTy <> " | " <> ppTrace boolTy)
  registerBuiltin BuiltinBoolFalse ctorFalse

registerIf :: (Members '[Builtins, NameIdGen] r) => FunctionDef -> Sem r ()
registerIf f = do
  bool_ <- getBuiltinName (getLoc f) BuiltinBool
  true_ <- toExpression <$> getBuiltinName (getLoc f) BuiltinBoolTrue
  false_ <- toExpression <$> getBuiltinName (getLoc f) BuiltinBoolFalse
  let if_ = f ^. funDefName
      u = ExpressionUniverse (Universe {_universeLevel = Nothing, _universeLoc = error "Universe with no location"})
  vart <- freshVar "t"
  vare <- freshVar "e"
  hole <- freshHole
  let e = toExpression vare
      exClauses :: [(Expression, Expression)]
      exClauses =
        [ (if_ @@ true_ @@ e @@ hole, e),
          (if_ @@ false_ @@ hole @@ e, e)
        ]
  registerFun
    FunInfo
      { _funInfoDef = f,
        _funInfoBuiltin = BuiltinBoolIf,
        _funInfoSignature = u <>--> bool_ --> vart --> vart --> vart,
        _funInfoClauses = exClauses,
        _funInfoFreeVars = [vare],
        _funInfoFreeTypeVars = [vart]
      }

registerOr :: Members '[Builtins, NameIdGen] r => FunctionDef -> Sem r ()
registerOr f = do
  bool_ <- getBuiltinName (getLoc f) BuiltinBool
  true_ <- toExpression <$> getBuiltinName (getLoc f) BuiltinBoolTrue
  false_ <- toExpression <$> getBuiltinName (getLoc f) BuiltinBoolFalse
  let or_ = f ^. funDefName
  vare <- freshVar "e"
  hole <- freshHole
  let e = toExpression vare
      exClauses :: [(Expression, Expression)]
      exClauses =
        [ (or_ @@ true_ @@ hole, true_),
          (or_ @@ false_ @@ e, e)
        ]
  registerFun
    FunInfo
      { _funInfoDef = f,
        _funInfoBuiltin = BuiltinBoolOr,
        _funInfoSignature = bool_ --> bool_ --> bool_,
        _funInfoClauses = exClauses,
        _funInfoFreeVars = [vare],
        _funInfoFreeTypeVars = []
      }

registerAnd :: Members '[Builtins, NameIdGen] r => FunctionDef -> Sem r ()
registerAnd f = do
  bool_ <- getBuiltinName (getLoc f) BuiltinBool
  true_ <- toExpression <$> getBuiltinName (getLoc f) BuiltinBoolTrue
  false_ <- toExpression <$> getBuiltinName (getLoc f) BuiltinBoolFalse
  let and_ = f ^. funDefName
  vare <- freshVar "e"
  hole <- freshHole
  let e = toExpression vare
      exClauses :: [(Expression, Expression)]
      exClauses =
        [ (and_ @@ true_ @@ e, e),
          (and_ @@ false_ @@ hole, false_)
        ]
  registerFun
    FunInfo
      { _funInfoDef = f,
        _funInfoBuiltin = BuiltinBoolAnd,
        _funInfoSignature = bool_ --> bool_ --> bool_,
        _funInfoClauses = exClauses,
        _funInfoFreeVars = [vare],
        _funInfoFreeTypeVars = []
      }

registerBoolPrint :: (Members '[Builtins] r) => AxiomDef -> Sem r ()
registerBoolPrint f = do
  bool_ <- getBuiltinName (getLoc f) BuiltinBool
  io <- getBuiltinName (getLoc f) BuiltinIO
  unless (f ^. axiomType === (bool_ --> io)) (error "Bool print has the wrong type signature")
  registerBuiltin BuiltinBoolPrint (f ^. axiomName)
