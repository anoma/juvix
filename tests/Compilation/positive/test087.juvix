-- Deriving Ord
module test087;

import Stdlib.Data.Fixity open;
import Stdlib.Data.Bool open;
import Stdlib.Data.Pair open;
import Stdlib.Trait.Ord open;
import Stdlib.System.IO open;

syntax alias compare := Ord.compare;

type T :=
  | One
  | Two
  | Three
  | Four;

open T using {One; Two; Three; Four} public;

deriving instance
ordT : Ord T;

type Tree (A : Type) :=
  | leaf
  | node (Tree A) A (Tree A);

open Tree using {leaf; node} public;

deriving instance
ordTree {A} {{Ord A}} : Ord (Tree A);

main : IO :=
  printLn (compare One One)
    >>> printLn (compare One Two)
    >>> printLn (compare Two One)
    >>> printLn (compare (One, One) (One, Three))
    >>> printLn (compare (Four, Four) (Four, Three))
    >>> printLn
      (compare (node leaf Four leaf) (node (node leaf One leaf) Two leaf))
    >>> printLn
      (compare
        (node (node leaf One leaf) Four leaf)
        (node (node leaf One leaf) Two leaf))
    >>> printLn
      (compare
        (node (node leaf One leaf) Four leaf)
        (node (node leaf One leaf) Four leaf));
