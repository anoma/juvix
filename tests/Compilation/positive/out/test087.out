Equal
LessThan
GreaterThan
LessThan
GreaterThan
LessThan
GreaterThan
Equal
